### ISSUE TITLE
make UI URls cleaner with less escaping

### ISSUE THREAD

**Issue by air** [mesosphere/marathon#2059]
From Marathon UI, if I click on my app `/dcos/dashboard` my browser navigates to https://marathon.mesosphere.com/#/apps/%2Fdcos%2Fdashboard, like this:

![image](https://cloud.githubusercontent.com/assets/18724/9398280/40d74d44-475a-11e5-8c9a-12de501f701f.png)

Can we use a non-escaped URL instead with `/` characters instead of `%2F`?
- The clean format would enable use of a [custom Chrome search engine](https://support.google.com/customsearch/answer/2630963?hl=en), e.g. `https://marathon.mesosphere.com/#/apps/%s`
- If I try the URL https://marathon.mesosphere.com/#/apps/dcos/dashboard I get a blank page:
  ![image](https://cloud.githubusercontent.com/assets/18724/9398243/bdd85302-4759-11e5-9df6-5363e3060cd2.png)

Implementation detail: We should accept double slashes after `apps` to allow for the case where you're scripting in a full app name. e.g. This is accepted `https://marathon.mesosphere.com/#/apps//dcos/dashboard`

**Comment by aldipower** [comment]
Hey @air.

This is unfortunately not possible. because  the '/' is a separator for the route itself. 
We have such URLs:
```
#/apps/%2Fsleep500/debug
#/apps/%2Fsleep500/configuration
#/apps/%2Fsleep50/sleep50.b442307b-475f-11e5-845f-56847afe9799
```

Where ```/sleep50.b442307b-475f-11e5-845f-56847afe9799``` is a task detail view or ```/configuration``` the configuration tab.

**Comment by philipnrmn** [comment]
I'd really like to see this working as well; we'd have to adapt the subroutes so that they didn't interfere.

One option would be to transpose it to the front, eg `#/debug/apps/sleep500`

Another would be to use a different separator eg `#/apps/sleep500!debug`

**Comment by aldipower** [comment]
The first option doesn't reflect the hierarchy of the pages and breadcrumbs.

The second option pretends that the appId would be ```sleep500``` instead of ```/sleep500```, because the route starts with a ```/```-```apps```.
Starting the route with a ```/``` is the behavior of the react-router itself and really can't be changed.
So the second option would be correctly written like this ```#/apps!/sleep500!debug```. Very ugly imho.. :) And what if you have a ```!``` in the appId itself?

But maybe we can find a solution with double slashes like supposed, if I think twice.
Instead of defining the router like this:
```
    <Route name="app" path="apps/:appId" handler={AppPageComponent} />
    <Route name="appView" path="apps/:appId/:view" handler={AppPageComponent} />
```
This could work too with some modifications on other places:
```
    <Route name="app" path="apps//:appId" handler={AppPageComponent} />
    <Route name="appView" path="apps//:appId/:view" handler={AppPageComponent} />
```

Then we could do this (API-calls looking like this too):
```
#/apps//sleep500/configuration
```

We have to test this.

**Comment by aldipower** [comment]
OK, my proposal isn't working, because an appId could have a lot of slashes, e.g. ```/group1/subgroup/applicationame``` . This would be ```#/apps//group1/subgroup/applicationame/configuration```. :(

**Comment by air** [comment]
We should be able to do what we want. It would be really disappointing to be blocked from a good URL scheme by a framework choice (react) : / 

I'd like to stick with slashes and not other characters like `!`.
As a user I just want to paste in my app ID (with or without the leading slash) and have it work.

**Comment by saturnboy** [comment]
Escaping app names makes sense, but escaping groups is pretty silly.  I can query the API like this:

**Comment by nnordrum** [comment]
I'd go a step beyond "I'd happily accept" and say, I think that / _should_ be disallowed as a valid character _because_ it's a separator--regardless of whether it makes cleaner URLs. I think that it does make cleaner URLs is a nice side effect, but we should be promoting best practices over allowing people to do whatever the hell they want, and putting a / in the names when we're using the / as a separator would NOTA be a best practice IMO. This this will affect somebody coming from some legacy tech debt situation, but they need to move forward. I got bit yesterday by the change in how dependencies are resolved, but I had to change because the change put us in a better position overall. 

TL;DR: +1 for removing / as a valid app name character :)

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-2870. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.

**Comment by meichstedt** [comment]
Note: This issue has been migrated to https://jira.mesosphere.com/browse/MARATHON-2870. For more information see https://groups.google.com/forum/#!topic/marathon-framework/khtvf-ifnp8.
